Source/Android/app/src/main/java/org/dolphinemu/dolphinemu/features/DocumentProvider.kt

```diff
@@ -0,0 +1,253 @@
+        var flags = 0
+        if (file.isDirectory && file.canWrite()) {
+            flags = DocumentsContract.Document.FLAG_DIR_SUPPORTS_CREATE
+        } else if (file.canWrite()) {
+            flags = DocumentsContract.Document.FLAG_SUPPORTS_WRITE
+            flags = flags or DocumentsContract.Document.FLAG_SUPPORTS_DELETE
+            flags = flags or DocumentsContract.Document.FLAG_SUPPORTS_REMOVE
+            flags = flags or DocumentsContract.Document.FLAG_SUPPORTS_MOVE
+            flags = flags or DocumentsContract.Document.FLAG_SUPPORTS_COPY
+            flags = flags or DocumentsContract.Document.FLAG_SUPPORTS_RENAME
+        }
+
+        val name = if (file == rootDirectory) {
+            context!!.getString(R.string.app_name)
+        } else {
+            file.name
+        }
+        cursor.newRow().apply {
+            add(DocumentsContract.Document.COLUMN_DOCUMENT_ID, pathToDocumentId(file))
+            add(DocumentsContract.Document.COLUMN_MIME_TYPE, getTypeForFile(file))
+            add(DocumentsContract.Document.COLUMN_DISPLAY_NAME, name)
+            add(DocumentsContract.Document.COLUMN_LAST_MODIFIED, file.lastModified())
+            add(DocumentsContract.Document.COLUMN_FLAGS, flags)
+            add(DocumentsContract.Document.COLUMN_SIZE, file.length())
+            if (file == rootDirectory) {
+                add(DocumentsContract.Document.COLUMN_ICON, R.drawable.ic_dolphin)
+            }
+        }
+    }
+
+    private fun pathToDocumentId(path: File): String {
+        return "$ROOT_ID/${path.toRelativeString(rootDirectory!!)}"
+    }
+
+    private fun documentIdToPath(documentId: String): File {
+        val file = File(rootDirectory, documentId.substring("$ROOT_ID/".lastIndex))
+        if (!file.exists()) {
+            throw FileNotFoundException("File $documentId does not exist.")
+        }
+        return file
+    }
+
+    private fun getTypeForFile(file: File): String {
+        return if (file.isDirectory)
+            DocumentsContract.Document.MIME_TYPE_DIR
+        else
+            MimeTypeMap.getSingleton().getMimeTypeFromExtension(file.extension)
+                ?: "application/octet-stream"
+    }
+
+    private fun findFileNameForNewFile(file: File): File {
+        if (!shouldReplaceFile(file)) {
+            return file
+        }
+
+        var unusedFile = file
+        var i = 1
+        while (unusedFile.exists()) {
+            val pathWithoutExtension = unusedFile.absolutePath.substringBeforeLast('.')
+            val extension = unusedFile.absolutePath.substringAfterLast('.')
+            unusedFile = File("$pathWithoutExtension.$i.$extension")
+            i++
+        }
+        return file
+    }
+
+    private fun shouldReplaceFile(file: File): Boolean {
+        val path = file.absolutePath
+        return !path.contains(WII_SAVE_DIR)
+                && !path.contains(GC_SAVE_DIR)
+                && !path.contains(SAVE_STATE_DIR)
```

Comment:
Fair, it's annoying that Android doesn't show a dialog to ask the user what to do in case there's a conflict.

So do we just always replace or always create new files?

Reply:
Always creating a new file feels like the safest option to me.

Reply:
That's safe but also probably annoying given that the most common use case is gonna be copy config files and texture packs.

Reply:
Okay, I changed it to create new files instead.